### ISSUE TITLE
[$16,000] Let's fix a React Native issue -- useWindowDimensions() returns swapped height and width in iOS

### ISSUE THREAD

**Issue by deetergp** [Expensify/App#2727]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

We have an issue where opening the app in native iOS on an iPad in Landscape orientation, the Chat Selector component is floating in the middle of the screen. We opened [this issue](https://github.com/Expensify/Expensify.cash/issues/2180) to deal with it and had a workaround put forward by one of our contributors, but in the process of investigating, discovered that the real issue is with React Native core. They have [an open issue](https://github.com/facebook/react-native/issues/29290) and are working on a fix, but it's anyone's guess when it will be ready.

Let's keep an eye on the React Native team's solution and see if we can put it into place and undo the the temporary workaround we are using for now.

cc @marcaaron @mallenexpensify 

## Workaround:
https://github.com/Expensify/Expensify.cash/issues/2180#issuecomment-833695795

## Original React Native Issue
https://github.com/facebook/react-native/issues/29290

## Platform:
Where is this issue occurring?

- [ ] Web
- [x] iOS (iPad)
- [ ] Android
- [ ] Desktop App
- [ ] Mobile Web

**Version Number:** 1.0.2-51
**Notes/Photos/Videos:** See the original issue

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days.  eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by mallenexpensify** [comment]
Removed the monthly label so it won't auto-close.  Also dropped a (likely fruitless) post in #expensify-open-source to see if anyone has ideas. https://expensify.slack.com/archives/C01GTK53T8Q/p1633992309160300

**Comment by puneetlath** [comment]
@deetergp do we know that they are actively working on a fix? If not, perhaps we can hire a contributor to submit a PR to fix it to RN directly now.

**Comment by deetergp** [comment]
> @deetergp do we know that they are actively working on a fix? If not, perhaps we can hire a contributor to submit a PR to fix it to RN directly now.

Your guess is as good as mine… The last comment on that issue was from a year ago. There are references from other issues and PRs as recently as the end of October '21 so it is definitely still in people's minds. 🤷

**Comment by MelvinBot** [comment]
@deetergp, this ***Monthly*** task hasn't been acted upon in 6 weeks; closing.

If you disagree, feel encouraged to reopen it -- but pick your least important issue to close instead.

**Comment by puneetlath** [comment]
I still think we should hire a contributor to fix this PR directly in RN https://github.com/facebook/react-native/issues/29290. I'm going to add the external label for it.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by NicMendonca** [comment]
Upwork job: https://www.upwork.com/jobs/~017b165dfa9821e51c

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcaaron (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by marcaaron** [comment]
Should we just quadruple this one now? Fixing this upstream is worth a bit more than $250.

**Comment by puneetlath** [comment]
Agreed.

**Comment by NicMendonca** [comment]
quadrupled the price -- https://www.upwork.com/jobs/~017b165dfa9821e51c

**Comment by marcaaron** [comment]
Waiting for proposals

**Comment by marcaaron** [comment]
Still waiting

**Comment by NicMendonca** [comment]
doubled: https://www.upwork.com/jobs/~017b165dfa9821e51c

**Comment by NicMendonca** [comment]
Doubled today ☝️

**Comment by marcaaron** [comment]
Still waiting for proposals

**Comment by mallenexpensify** [comment]
Doubled price to $4000
https://www.upwork.com/jobs/~017b165dfa9821e51c

**Comment by marcaaron** [comment]
Still waiting for proposals here.

**Comment by mallenexpensify** [comment]
Doubled price to $8000

**Comment by aneequeahmad** [comment]
### PROPOSAL

As the problem is already highlighted in the react-native community as this is a react-native bug. 

**Solution 01**

using the `useSafeAreaFrame()` hook from `react-native-safe-area-context`. It works really well as a replacement if you make sure to only have one SafeAreaProvider at the top level of your app. Many libraries like `react-native-collapsible-tab-view` have done this to address this `useWIndowDimensions()` issue.

**Solution 02**

Another way to achieve what we want would be to allow passing the width as an optional prop while still defaulting it internally to what `useWindowDimensions` returns.
Pass our own `width` which you can get from `useSafeAreaFrame`. Additionally, it will also allow the library to work properly 
If you could look into this alternate implementation.

In my opinion, both of these solutions are bulletproof and as is not a workaround but a proper alternate solution.

**Comment by marcaaron** [comment]
>works really well as a replacement

Sorry not looking for replacements (we already have a workaround deployed). 
Fix in react-native repro is a requirement for this issue.

**Comment by lbaldy** [comment]
PROPOSAL
This has to be fixed inside the react native repository, through a standard MR process for Facebook (out of our control in terms of timing). Once merged, it would also be required to upgrade Expensify's react native version (this should probably be handled via another ticket that is well planned etc.).

Solution
I debugged the react-native code, and to me it seems that react native publishes the orientation change event one extra time when switching the state of the app to 'inactive'. Here is what is happening:

1. iPad is in landscape.
2. We move the app to inactive state.
3. Native Code queues portrait orientation change (no such change happened), and immediately after it triggers landscape change (same as we had in point 1).
4. We restore the app to active state.
5. The app receives two queued orientation change events, one after another.
6. The quick transition between portrait and landscape happens even though it never went back to portrait.

IMHO solution should be applied to react native core modules, preventing it from publishing this event in case of app state change.

Please let me know if the proposal is fine from your perspective, and I will work on a react native MR.

I can provide a quickfix to react native code if you would like to test if it really fixes the issue.

Obviously before all the tests are going to happen the workaround implemented for this has to be removed.

**Comment by AndrewGable** [comment]
Have we confirmed this happens on a "basic" react native app to rule out any UI logic on our side? I know that it seems like they were "flipped", but it seems possible that the values were just "delayed" in reporting the real value?

**Comment by lbaldy** [comment]
@AndrewGable Please find my video showing the issue, and reproduction using the react native basic application.

https://youtu.be/nFDOml9M8w4

**Comment by AndrewGable** [comment]
Thank you for the very informative video @lbaldy! This looks very promising for sure. @parasharrajat - Can you review the videos and verify this is the expected behavior then work with @lbaldy to flush out the proposal?

**Comment by parasharrajat** [comment]
Oh, I didn't notice that I am assigned to it. I will check them tomorrow morning IST.

**Comment by lbaldy** [comment]
@azimgd  that sounds exactly like what I presented in my proposal and a follow up video to @AndrewGable. No? ;)

**Comment by parasharrajat** [comment]
Looking at the video @lbaldy I can say that it looks convincing. The next step would be to share a technical proposal here to explain the problem and lay out your fix. Your previous solution does not really talk about the solution but what is happening. 

Also, the solution should be tested on the E/app without the workaround before being proposed here. 

--- 
And, yes it is necessary to explain your technical changes here in the proposal.

**Comment by azimgd** [comment]
@marcaaron @parasharrajat could you have a look at the proposal above please ?

**Comment by lbaldy** [comment]
Problem statement:

The orientation change notification is published by IOS when app goes to the background, due to some sort of a race condition/timing issue causes wrong dimensions being returned and transformation being queued in react native.

Solution description and assumptions:

We need to prevent publishing the event when app transforms from foreground to background, but at the same time we have to ensure:

1. App reacts to orientation changes when in foreground.
2. App reacts to orientation changes that happened when the app was in background.
3. App reacts to view changes when in multitasking (split) view.
4. App reacts to view changes when moving from slide over view to fullscreen or split.
5. App doesn't re-trigger the animation of the side bar when app simply goes to background and comes back to foreground in the same orientation.

Technical solution:

The fix has to be applied directly to react-native core module. The whole code responsible for this behavior lives in RCTDeviceInfo.mm.

There are a couple of things that have to be modified.

a) The body of the \_interfaceOrientationDidChange:

Here we make sure we run this function when the app is in active and also it was switching to fullscreeen or is not running in fullscreen.

```
- (void)_interfaceOrientationDidChange
{
  UIInterfaceOrientation nextOrientation = [RCTSharedApplication() statusBarOrientation];
  UIApplicationState appState = [RCTSharedApplication() applicationState];
  BOOL isRunningInFullScreen = CGRectEqualToRect([UIApplication sharedApplication].delegate.window.frame, [UIApplication sharedApplication].delegate.window.screen.bounds);

  // Update when we go from portrait to landscape, or landscape to portrait
  if ((((UIInterfaceOrientationIsPortrait(_currentInterfaceOrientation) &&
       !UIInterfaceOrientationIsPortrait(nextOrientation)) ||
      (UIInterfaceOrientationIsLandscape(_currentInterfaceOrientation) &&
       !UIInterfaceOrientationIsLandscape(nextOrientation)) || (isRunningInFullScreen != _isFullscreen || !isRunningInFullScreen)))
      && appState == UIApplicationStateActive) {
      #pragma clang diagnostic push
      #pragma clang diagnostic ignored "-Wdeprecated-declarations"
          [[_moduleRegistry moduleForName:"EventDispatcher"] sendDeviceEventWithName:@"didUpdateDimensions"
                                                                                body:RCTExportedDimensions(_moduleRegistry)];
            _currentInterfaceOrientation = nextOrientation;
            _isFullscreen = isRunningInFullScreen;
      #pragma clang diagnostic pop
  }


}
```

b) \_interfaceFrameDidChange

Here we make sure we run it when the app is in active state.

```
- (void)_interfaceFrameDidChange
{
  NSDictionary *nextInterfaceDimensions = RCTExportedDimensions(_moduleRegistry);
    UIApplicationState appState = [RCTSharedApplication() applicationState];

  if (!([nextInterfaceDimensions isEqual:_currentInterfaceDimensions]) && appState == UIApplicationStateActive) {
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
        [[_moduleRegistry moduleForName:"EventDispatcher"] sendDeviceEventWithName:@"didUpdateDimensions"
                                                                              body:nextInterfaceDimensions];
          _currentInterfaceDimensions = nextInterfaceDimensions;
    #pragma clang diagnostic pop
  }
}
```

c) Add the \_isFullscreen variable to hold the previous state for multitasking view.

```
@implementation RCTDeviceInfo {
  UIInterfaceOrientation _currentInterfaceOrientation;
  NSDictionary *_currentInterfaceDimensions;
  BOOL _isFullscreen;
}
```

d) We need to make sure to run the interfaceOrientationDidChange instead of interfaceFrameDidChange when the UIApplicationDidBecomeActiveNotification occurs.

```
  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(interfaceOrientationDidChange)
                                               name:UIApplicationDidBecomeActiveNotification
                                             object:nil];
```

Solution provided by azimgd doesn't work when the app changes the orientation while in background.

**Comment by parasharrajat** [comment]
> App reacts to orientation changes that happened when the app was in the background.

How can an app change orientation while in the background? Could you please explain me? @lbaldy

@azimgd Why are we removing the _bridge parameter from function definitions?

**Comment by parasharrajat** [comment]
Got it. Why are you only triggering the event when the app is active? https://github.com/Expensify/react-native/blob/f635dd7b74b9d5f2d3b71972bcc395eae780ed6d/React/CoreModules/RCTDeviceInfo.mm#L213

**Comment by azimgd** [comment]
When the app is not in split screen mode:
- Trigger when the app is opened (goes from background → foreground)
- Trigger when the device is rotated

We are filtering away wrong dimensions emits (e.g. goes from foreground → background) there. As when the app goes into background those are triggered in batch [wrong{x,y}, correct{x,y}].

Additionally `!isAppFullscreen` will:
- trigger when the app is resized in the split screen mode

**Comment by lbaldy** [comment]
@parasharrajat I added a video testing all the cases as well as 'background orientation change' I mentioned earlier: https://youtu.be/Ytj0K4SwP4w

**Comment by parasharrajat** [comment]
Thanks, I am testing something and soon update you all.

**Comment by parasharrajat** [comment]
Ok, based on my testing. There is no event fired when the app is in the background. `background orientation change` or when the app's orientation is changed while the app is in an inactive state, is detected when the app comes to the active state. As soon as the app comes to an active state, the app receives an orientation event with proper data. 

--- 

That said, @lbaldy came up with the proposal first. We were mostly convinced by his explanation and videos. He was also thorough during the discussion. Although @azimgd's proposal is also correct it matches @lbaldy 's proposal. Given that @lbaldy is a new contributor and he mentioned hiding the code "For obvious reasons I blurred the XCODE part of the video.", I asked him to share technical details. I like @lbaldy 's proposal. @lbaldy Please consider sharing the technical details next time. It is a crucial part of the proposal review process. 

It was very hard to pick one over another due to both proposals were great, I chose to go with the first proposal. 

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by azimgd** [comment]
Based on the `contributing.md`:
> We look for the earliest provided, best proposed solution that addresses the job.

That said:
- previous proposals didn't have any technical explanation
- missed a very important "Split Screen Mode", which is a key on why duplicate events happen
- previous proposals mentioned a "one-liner fix" which was later transformed into very close copy of [PR#9](https://github.com/Expensify/react-native/pull/9), see `isFullScreen`, event filtering

> convinced by his explanation and videos
> For obvious reasons I blurred the XCODE part of the video
> It is a crucial part of the proposal review process.

`contributing.md` explicitly mentions: Your solution proposal should include a brief technical explanation of the changes you will make.

**Comment by lbaldy** [comment]
I think the solution proposed by @azimgd  doesn't work or even breaks when the app when in example: switches from slide out mode to fullscreen. Thus this proposal isn't complete.

**Comment by parasharrajat** [comment]
Thanks to both of you for sharing your concerns. But my choice is never final. There would be someone from the Team to assign the job to one of you. We are currently discussing this internally.

**Comment by azimgd** [comment]
> doesn't work or even breaks when the app when in example: switches from slide over mode to fullscreen.
Have you had a chance to test my PR? I'm pretty sure it handles that scenario correctly.

**Comment by lbaldy** [comment]
@parasharrajat I am going to wait for the final decision and assignment after the internal discussions. If any additional information regarding any of my comments or videos or code is needed please let me know.

**Comment by marcaaron** [comment]
Sorry guys, going OOO for a week and won't be able to help with this one. Sounds like it's moving in the right direction though!

**Comment by mallenexpensify** [comment]
Quick update before the weekend hits, we're still discussing, expect an update on Monday, thanks for the patience and the help so far

**Comment by mallenexpensify** [comment]
Thanks for the patience, we're actively discussing and making updates that will help with issues like this where the final deliverable is to have a PR merged in a 3rd party repo. 

We made an [update to the CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#propose-a-solution-for-the-job) 
> Solution: Update the CONTRIBUTING.md to include `written` "Your solution proposal should include a brief **written** technical explanation of the changes you will make""

@lbaldy , your videos were helpful and easy to follow, the reason we added `written` is because it makes it easier to compare proposals to ensure "Any new proposal should be substantively different from existing proposals." 

We hope to have another update by EOW (I'm off the next two days so it might not be til next Monday).

**Comment by NicMendonca** [comment]
@mallenexpensify any update by chance?

**Comment by mallenexpensify** [comment]
Apologies for the delay.  I drafted a reply with details, I need to get more internal 👀 on it before posting since this is a unique edge case and because we're also wanting to create a process for third-party repos.  I'll follow up in the morn.

**Comment by mallenexpensify** [comment]
First, a paraphrased plan for how we plan to compensate for issues that require fixes in third party repos.
**tl,dr:** (and this isn't set in _stone_ until the .md file is created)
- Half will be paid when the fix is merged in our fork, the E/App repo is pointed at the fork and it doesn't break for a week. 
- The other half will be paid when the fix is merged upstream, fork is update w/ upstream fix, everything _works_ for a week . 
 
If there’s not a fork, payment would be issued once a fix is merged upstream and released, E/App is updated to use the new dependency version and nothing breaks for a week. 

For this specific issue, technically @lbaldy was first and there’s not (apparently) a substantial difference in the following written proposal by @azimgd.  We have since updated our CONTRIBUTING.md to require written proposals in the future to avoid similar circumstances.   

I propose we hire @lbaldy  for $16k to fix the issue with half the payment due once the PR is merged in our fork, the E/App repo is pointed at the fork and there are no regressions for a week. The remaining $8k will be paid to @lbaldy when the fix is merged upstream, fork is updated w/ upstream fix and there are no regressions for a week. 

Our intent is to get the PR merged upstream so we want to ensure contributors have an incentive to do so.  For @azimgd ,  compensation will be $8k due once the PR is merged upstream (and no regressions for a week). The deliverable for @azimgd  is to follow along as the work is being done on the PR for E/App repo to ensure the final PR is (hopefully) likely to be merged in the 3rd party repo. Then, both @lbaldy and @azimgd  will push (with help from others) to get @lbaldy  ’s PR merged into the 3rd party repo. 

Since this is new for us please take notes on what works and how things might be improved for getting PRs merged in 3rd party repos.  Please share those notes in this GH and/or #expensify-open-source.

**Comment by lbaldy** [comment]
Thanks @mallenexpensify  for the explanation and IMHO great way of resolving this. What you described sounds good to me.

I just have one additional comment to the update you did here:

> We made an [update to the CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#propose-a-solution-for-the-job)
>Solution: Update the CONTRIBUTING.md to include written "Your solution proposal should include a brief written technical explanation of the changes you will make""

I think it would be worth to mention that the technical explanation means code changes (what files, what lines to add/remove/change).

Something like this:

>Your solution proposal should include a brief written technical explanation of the changes you will make, **like files to change, critical pieces of code required to present the high level idea of the change.**

### PR

In terms of the PR (https://github.com/Expensify/react-native/pull/10).

I think the commit in this PR isn't signed. 

I don't want to touch it until I get a green light - when I get one I'll do a force push to it so the commit becomes a signed commit.

Also, when the change is merged to a react-native fork, the Expensify/App will use the new version, but also the existing workaround fix has to be removed (https://github.com/Expensify/App/issues/2180#issuecomment-833695795). Shall I open a PR for this one as well or will that be handled internally while updating App's version of react native? 

### Payment

In terms of splitting the payment into 2 that sounds good to me, only concern is that it falls into higher Upwork commission when it is below 10k. I am thinking if it makes sense to simply do a single payment once it is merged upstream to Facebook's repo. I hope we can discuss that later when we actually prepare a plan.

### Merge upstream

In terms of merging the code to Facebook's react native repo. The PR requester has to sign the https://github.com/facebook/react-native/wiki/Contributor-License-Agreement. I have signed it because I did a PR to their repo in the past. I am not sure how that works when it is a merge from another fork that I am not the owner of - https://github.com/Expensify/react-native. I am just wondering if anyone from Expensify has to adhere/sign the CLA or we just file a new PR to Facebook's repo standalone. Something to flesh out I guess.

What should be the next step?

**Comment by mallenexpensify** [comment]
@lbaldy I'm about to head out, @parasharrajat will followup soon then I will too next week.  
For the CONTRIBUTING.md update, that makes sense, I want to run by folks internally then can do it next week. 

For the payment commissions, can you double check? We're Enterprise and I thought contributors/freelancers paid a fixed rate (10% was  my memory), regardless of $ amount.

**Comment by azimgd** [comment]
Looks like this issue has been fixed on a: `"react-native": "0.68.2"`. @parasharrajat could you double check and verify both android and ios versions ?

**Comment by parasharrajat** [comment]
Oh. Great. Thanks I will check it.

**Comment by lbaldy** [comment]
@azimgd  is there any particular commit in react native repository you're referring to?

I just checked the same scenario I used in the original 'react-native' repro (0.68.1). And I still observe the same behaviour - two events (portrait, landscape) when bringing the app to background. I also checked their changelog.md, I don't see anything that could seem related either. Also, checked the DeviceInfo.mm file, and still don't see anything similar to the fixes proposed. Is there anything else you're referring to?

The dependencies used in my repro project:

`
  "dependencies": {
    "react": "17.0.2",
    "react-native": "0.68.2"
  },
`

**Comment by parasharrajat** [comment]
> Also, when the change is merged to a react-native fork, the Expensify/App will use the new version, but also the existing workaround fix has to be removed (https://github.com/Expensify/App/issues/2180#issuecomment-833695795). Shall I open a PR for this one as well or will that be handled internally while updating the App's version of react-native?

yes, Open a new PR with an updated fork on react-native. 

>  it falls into higher Upwork commission when

Matt is correct here. 

So you can create a new PR from our fork to react-native. That should be possible and as you will be the author, we probably don't have to do anything.

**Comment by mallenexpensify** [comment]
You betcha @lbaldy , please apply here and confirm in comment once you have
https://www.upwork.com/jobs/~01411bf531c59e797e

**Comment by lbaldy** [comment]
Thanks @mallenexpensify. Applied.

**Comment by imran-ranglerz** [comment]
const isPortrait = () => {
    const dim = Dimensions.get('screen');
    return dim.height >= dim.width;
};
 
/**
 * Returns true of the screen is in landscape mode
 */
const isLandscape = () => {
    const dim = Dimensions.get('screen');
    return dim.width >= dim.height;
};

by using this we can find out the dimension and solve this.

**Comment by mallenexpensify** [comment]
@lbaldy , hired you in Upwork, assuming you're `Lukasz B` there

@imran-ranglerz , we already have hired someone for this job, to check all open jobs, click https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22

**Comment by lbaldy** [comment]
@mallenexpensify  thanks. That me, I've received the Upwork email too. The PR is already opened and I think @parasharrajat is on it.

**Comment by parasharrajat** [comment]
Expect an update soon. Just too exhausted today 😴🛌

**Comment by parasharrajat** [comment]
I will start reviewing this tomorrow daytime IST. Overall the PR looks good to me. 
@lbaldy Have you created the issue and PR on RN yet?

**Comment by lbaldy** [comment]
I haven't as per:

> First PR to our RN fork.
> Second PR to this repo to update the fork and remove the Hack.
> Then another PR to react-native.

Assuming we should merge it first to Expensify fork, test if all works well for 7 days, and then pull it into react native repo.

Please let me know if you see it different, and I will go ahead and fire a request for Facebook.

**Comment by parasharrajat** [comment]
Ok, that works for.

**Comment by mallenexpensify** [comment]
>Assuming we should merge it first to Expensify fork, test if all works well for 7 days, and then pull it into react native repo.

Good plan!

**Comment by NicMendonca** [comment]
@lbaldy @parasharrajat any updates here?

**Comment by parasharrajat** [comment]
Fork PR is ready to be merged. 

Completion: 30%
- [x] PR on the Expensify fork of RN. 
- [ ] PR for upgrading the version of RN on E/app.
- [ ] PR to RN.

**Comment by lbaldy** [comment]
The MR has been merged to master. Have you ever used react native fork in your E/App or have you always been using the official react native release from npm?

**Comment by parasharrajat** [comment]
I think we always used the official package. And we try use forked repo with V0.68, it may break stuff.

**Comment by lbaldy** [comment]
Okay so there are a couple of things we need to flesh out.

a) Upgrading a react native version for an app is always a big project (dev work as well as regression check across the whole app) do we really want to do it twice (once to fork version once to facebooks version)? 
b) E/App is not and has not been using the for ever, why do we want to use fork now instead of just getting it straight from official repo?
c) Should the react native version upgrade be handled as a 'side effect' of a bugfix?

**Comment by lbaldy** [comment]
Anyway, I am working on getting the forked version to work. Once ready I will update the https://github.com/Expensify/App/pull/9143

**Comment by parasharrajat** [comment]
We are already planning to upgrade the app to latest RN 0.68. That should help in transition. https://github.com/Expensify/App/issues/8503#issuecomment-1152785723. But it is still in planning phase. I am not sure if someone is already working on that. 

@mallenexpensify Would you be able to forward the discussion on the other issue? Let's see if that can be handled. Thanks.

**Comment by mallenexpensify** [comment]
> @mallenexpensify Would you be able to forward the discussion on the other issue? Let's see if that can be handled. Thanks.

@parasharrajat , are you referring to the _new_ plan for how to manage third-party repo issues?

**Comment by parasharrajat** [comment]
No, I meant about this issue https://github.com/Expensify/App/issues/8503#issuecomment-1152785723.

**Comment by lbaldy** [comment]
No matter what we do with the upgrade, I think it makes sense to open a Merge Request to Facebook already to keep the ball rolling.

@parasharrajat do you agree or do you think we should wait for our 7-day period test in E/App before we go to FB?

**Comment by parasharrajat** [comment]
> I will start reviewing this tomorrow daytime IST. Overall the PR looks good to me. 
> @lbaldy Have you created the issue and PR on RN yet? 

That's what I was suggesting here.

**Comment by lbaldy** [comment]
I opened a PR for Facebook: https://github.com/facebook/react-native/pull/34014

**Comment by parasharrajat** [comment]
@lbaldy Let's add the reproduction repo to the PR showing the real issue. You can also reuse samples from old issues.

**Comment by lbaldy** [comment]
@parasharrajat  thanks for the inputs. Updated the MR, added repro the PR is now following the required format. I guess we now have to wait.

**Comment by lbaldy** [comment]
Update: 

The MR for Facebook (https://github.com/facebook/react-native/pull/34014 has) has been 'imported', this means it is running though their internal checks at Meta. There are two tasks at Meta that are failing but external contributors don't seem to have access to any details about that.

On top of that, I worked upgrading to the 'main' version from Expensify's fork. This isn't trivial, this is not just a react native upgrade to 0.68.x but that's way more. Master is ahead of 0.68.x obviously, moreover it requires react update to 18.0.0. And master isn't stable obviously, we should not be doing that imho.

I was wondering if we can do something to get the fix into E/App as quickly as possible - I see a couple of options here. 

a) We take the Expensify/react-native fork, we checkout the tag equal to the current version of react native used in E/App -> and we apply our fix there. Then we can simply switch the react native to use the forked version. - This sounds pretty safe and simple. We may need to do some magic to update android along with ios - but that's well documented.

b) We apply a monkey patch. Simply by adding a npm postinstall script, we monkey patch the RCTDeviceInfo.mm file - this is a bit hackish.

c) We wait for the PR merged and we perform a full upgrade - this brings all the risks and efforts of a full upgrade as well as waiting time to get the fix in.

**Comment by parasharrajat** [comment]
a works for me.

**Comment by lbaldy** [comment]
@parasharrajat 

Here is a working version: https://github.com/lbaldy/ExpensifyApp/tree/bugfix/2727-upgrade-to-fork

Which uses this fork branch: https://github.com/lbaldy/expensify-fork-react-native/tree/bugfix/2727_on_0.66.4

I cannot open a PR, because we don't really want to merge it to E/react-native#main.

Ideally, if you could create a branch inside E/react-native from the v0.66.4 tag, after that I'd be able to fire a PR from my branch to your branch.

Once the above is done, I'll open a MR (lbaldy/expensify-fork-react-native/tree/bugfix/2727_on_0.66.4 to E/react-native#0.66.4_patch). Once this is merged, I will update the lbaldy/ExpensifyApp/tree/bugfix/2727-upgrade-to-fork to use your patched version. And only after that will open a final MR to get the E/App#main fixed via the fork.

Sorry for so much back and forth, but this seems to be the only way to get this in with minimal impact on unrelated areas.

**Comment by parasharrajat** [comment]
Thanks for the suggestion @AndrewGable. I am also in favor of that if we are fine with the delay. We can handle this via https://github.com/Expensify/App/issues/8503#issuecomment-1152785723 https://github.com/Expensify/App/issues/2727#issuecomment-1154431164

**Comment by lbaldy** [comment]
I come bearing good news, that the PR https://github.com/facebook/react-native/pull/34014 has been merged to react-native's main branch! :)

**Comment by parasharrajat** [comment]
Debouncing the DimensionsChange event https://github.com/Expensify/App/blob/02f433979a7d22baa7d544ba5b6e236d77e7e0e3/src/components/withWindowDimensions.js#L42.

**Comment by lbaldy** [comment]
> Good news. Let me open an internal discussion to understand the timeline and next step.

Great. Please let me know about the outcome whenever you get to any consensus.

**Comment by parasharrajat** [comment]
Yeah, it is in progress.

**Comment by parasharrajat** [comment]
Still discussing internally...

**Comment by parasharrajat** [comment]
@stephanieelliott https://expensify.slack.com/archives/C02NK2DQWUX/p1656446682985319

**Comment by iwiznia** [comment]
No update

**Comment by lbaldy** [comment]
Hey @iwiznia, @mallenexpensify , @parasharrajat it seems that the change will eventually land in the react native 0.70.  A react native [0.70.0-rc.0](https://github.com/facebook/react-native/releases/tag/v0.70.0-rc.0) is available and it contains the change. I've run a basic `react-native init latestreact --version=0.70.0-rc.0` and the downloaded npm package had my changes.

**Comment by parasharrajat** [comment]
We are looking forward to upgrading the app to the latest RN version. @lbaldy do share your updates as well.

**Comment by lbaldy** [comment]
As I mentioned in my previous comment I am monitoring situation on React Native release. Seems like this change is going to land in 0.70.x which is currently in RC.0 state.

@parasharrajat I know there are multiple issues that talk about react native version upgrade. Has this been decided in the internal discussions if the upgrade is supposed to happen under this issue, some other issue or internally? Basically if I am the person that is supposed to do the upgrade of the whole app alone?

**Comment by parasharrajat** [comment]
Thanks for the update. I will let you know about this. we previously had discussion on the same but there were no conclusions.

**Comment by lbaldy** [comment]
Sure, thanks. I think we still have some time till 0.70.x gets official, also React Native often has some follow up minor releases right after, with critical fixes (that seems to be the case with 0.69.x).

**Comment by parasharrajat** [comment]
I expect at least a month for that.

**Comment by iwiznia** [comment]
We are waiting for a new release of react native

**Comment by lbaldy** [comment]
I see the #9841 is merged. I will build the app and see if my change is in and works good.

**Comment by lbaldy** [comment]
Hi @parasharrajat.

I just checked the latest code base. I run the latest https://github.com/Expensify/App . I managed to run the `npm ci `properly and it installed all the deps, when I look into `node_modules/react-native` I can see it installed the `@expensify/react-native` version `0.69.4`

`  "name": "@expensify/react-native",
  "version": "0.69.4",`

however this version does not seem to contain my fix in `node_modules/react-native/React/CoreModules/RCTDeviceInfo.mm` whereas it is present on the main branch here: https://github.com/Expensify/react-native/blob/main/React/CoreModules/RCTDeviceInfo.mm

Is the @expensify/react-native#0.69.4 same as Metas react-native#0.69.4 without any fork specific changes?

**Comment by mallenexpensify** [comment]
Hi, coming back from OOO, @parasharrajat , can you provide an update on where we're at? Thanks

**Comment by mallenexpensify** [comment]
Thanks for the detailed update @parasharrajat ! Gonna leave this a weekly, even though it might take a bit for RN to update, want to make sure to get folks paid timely (when the time comes...)

**Comment by mallenexpensify** [comment]
Still waiting...

**Comment by lbaldy** [comment]
The React Native 0.70.0 is finally out!

https://github.com/facebook/react-native/releases/tag/v0.70.0

I will work on a branch with an update and keep you posted in this MR.

**Comment by lbaldy** [comment]
I think your fork if missing latest tags. When I run:

`git tag | grep v0.70.0`

It shows nothing, react native CLI uses tags to identify the latest content.

**Comment by lbaldy** [comment]
@parasharrajat  I have updated  the PR we had: https://github.com/Expensify/App/pull/9143

This one uses the official React Native release 0.70.0 from Facebook. The fix is there and it seems to be working just fine.

The Expensify NPM repo doesn't contain the latest version yet - https://www.npmjs.com/package/@expensify/react-native.

If we wanna go with the fork, someone from Expensify needs to publish a new NPM package, once done I can switch my PR to use the forked version.

**Comment by lbaldy** [comment]
Great, keep me posted.

**Comment by iwiznia** [comment]
I think we are still battling with the fork

**Comment by lbaldy** [comment]
@parasharrajat  thanks for the update - my PR has been updated https://github.com/Expensify/App/pull/9143 - synced with main and revert of #29290 applied.

I run a quick smoke test, the React Native fix is there in node_modules as well as the fix seems to be working well.

**Comment by lbaldy** [comment]
Done, all updated including the videos.

**Comment by mallenexpensify** [comment]
@parasharrajat @iwiznia 👀 plz....
From [this comment](https://github.com/Expensify/App/issues/2727#issuecomment-1132353347), 
> I propose we hire @lbaldy for $16k to fix the issue with half the payment due once the PR is merged in our fork, the E/App repo is pointed at the fork and there are no regressions for a week. The remaining $8k will be paid to @lbaldy when the fix is merged upstream, fork is updated w/ upstream fix and there are no regressions for a week.
>
> Our intent is to get the PR merged upstream so we want to ensure contributors have an incentive to do so. For @azimgd , compensation will be $8k due once the PR is merged upstream (and no regressions for a week). The deliverable for @azimgd is to follow along as the work is being done on the PR for E/App repo to ensure the final PR is (hopefully) likely to be merged in the 3rd party repo. Then, both @lbaldy and @azimgd will push (with help from others) to get @lbaldy ’s PR merged into the 3rd party repo.

The PR was merged upstream a while ago, right? [here](https://github.com/facebook/react-native/pull/34014), then [The React Native 0.70.0 is finally out!](https://github.com/Expensify/App/issues/2727#issuecomment-1239718792)

1. So... @lbaldy and @azimgd are due compensation of $8k each, correct? 
2. And.. it appears (🤞) that @lbaldy will be due the other $8k soon, is that correct?

**Comment by iwiznia** [comment]
That looks right , @rajat and @lbaldy, can you confirm that the fix is in the RN repo (not our fork) and that all is done?

**Comment by parasharrajat** [comment]
@iwiznia That is @parasharrajat not @ rajat.

**Comment by parasharrajat** [comment]
Before we do that, there is a PR https://github.com/Expensify/App/pull/9143 to remove the workaround code from our app. I am reviewing it now and get it merged soon.

**Comment by iwiznia** [comment]
> @iwiznia That is @parasharrajat not @ rajat.

Sorry!

**Comment by mallenexpensify** [comment]
I paid @azimgd $8k because I was supposed to be going OOO for a while and I wanted to get it wrapped up. 
https://www.upwork.com/jobs/~01e51c349e0ee2b5c1

@lbaldy , I looked into paying you $8k as well but since the job we hired you for is 'fixed price', the only option for partial payment appeared to be to change the job price.  Which... we _could_ do then create another for $8k, I didn't want to do that before addressing with you.

**Comment by lbaldy** [comment]
@mallenexpensify  I am fine getting paid whole amount afterwards. Whichever is simpler and less hassle for you.

**Comment by lbaldy** [comment]
Hi @parasharrajat. Now, once the https://github.com/Expensify/App/pull/9143 was merged, what should be the next step?

**Comment by parasharrajat** [comment]
7 days after PR is in Production and there are no regressions, payment will be released and we will close this chapter.

**Comment by mallenexpensify** [comment]
Since this is a big payday, @iwiznia and/or @parasharrajat , can you confirm
- @lbaldy = $16k for job - https://www.upwork.com/jobs/~01411bf531c59e797e  
- @parasharrajat = $16k for C+ - New job here https://www.upwork.com/jobs/~01789a28c4422d1d15, please apply. 
- no one for reporting. 

Look/sound right?

**Comment by mallenexpensify** [comment]
Paid @lbaldy $16,000!  Thanks for the help!  Waiting to pay @parasharrajat so leaving open for now, bumped back to weekly.

**Comment by lbaldy** [comment]
@mallenexpensify  got that, thanks all for cooperation!

**Comment by mallenexpensify** [comment]
@parasharrajat , can you please accept so I can pay you and close this out?
https://www.upwork.com/jobs/~01789a28c4422d1d15

**Comment by mallenexpensify** [comment]
@parasharrajat please accept the job and @joekaufmanexpensify will pay you out (I'm heading OOO)  https://www.upwork.com/jobs/~01789a28c4422d1d15

**Comment by parasharrajat** [comment]
I will accept it in 2-3 days.

**Comment by joekaufmanexpensify** [comment]
Just assigned this issue, and it seems like all that needs to be done here is pay @parasharrajat as soon as they accept the job.

**Comment by joekaufmanexpensify** [comment]
Job still needs to be accepted before this can be paid.

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
@parasharrajat Any update on accepting this so we can pay you?

**Comment by parasharrajat** [comment]
@joekaufmanexpensify Sent a proposal.

**Comment by joekaufmanexpensify** [comment]
Great, thanks! Hired.

**Comment by joekaufmanexpensify** [comment]
@parasharrajat payment sent and contract ended.

**Comment by joekaufmanexpensify** [comment]
Closed job in upwork. Closing as payment has been issued, and this is all set!
